### PR TITLE
Implement associated consts for traits

### PR DESCRIPTION
**Pull request by tritao**
## Description

This is the last step to getting associated consts to work.

First it refactors the existing const support to use a newly-introduced `ConstantExpression` instead of relying on `VariableExpression` like we've been doing previously.

Then it adds IR generation for associated constants and enables the relevant tests.

[Add ConstantExpression and IR generation support.](https://github.com/FuelLabs/sway/commit/3d2e62598fec4955d06a1e2b4bd2fa8a2821af02)

[Update ConstantExpression constant before IR generation.](https://github.com/FuelLabs/sway/commit/4ec7ebe807348528c137d75ac50e3ab562de63af)

[Add IR generation for associated consts.](https://github.com/FuelLabs/sway/commit/67386a874e7eada3d47b161c4fa305ebca85193f)

Then there also some commits adding some new documentation:

[Add traits documentation to the reference.](https://github.com/FuelLabs/sway/pull/4301/commits/afda9026a0f66b981ad11b8d8a66b37855d712aa)

[Add some documentation about associated consts to the book.](https://github.com/FuelLabs/sway/pull/4301/commits/6845ddc1662d5ddece4d922cf60fb5996f8ddbc3)

Closes https://github.com/FuelLabs/sway/issues/3797.

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: docs/reference/src/documentation/language/built-ins/constants.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Constants
+
+A constant in Sway is an `associated item` which is an item associated to a particular struct type.
+
```
Comment (otrho) [inline docs/reference/src/documentation/language/built-ins/constants.md:4]: What about global `const FOO: bool = false;` style constants?  Are they documented elsewhere?

File: docs/reference/src/documentation/language/traits/index.md [hunk 2]
```diff
@@ -1 +1,49 @@
...
+All traits define an implicit type parameter `Self` that refers to "the type that is implementing this interface".
+Traits may also contain additional type parameters. These type parameters, including `Self`, may be constrained by
+other traits and so forth as usual.
+
+Traits are implemented for specific types through separate implementations.
+
+## Associated functions
+
+Trait functions consist of just a function signature, so they have to omit the function body by
+replacing it with a semicolon. This indicates that the implementation must define the function.
```
Comment (otrho) [inline docs/reference/src/documentation/language/traits/index.md:30]:
```suggestion
Trait functions consist of just a function signature. This indicates that the implementation must define the function.
```

File: docs/book/src/basics/consts.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+- The `mut` keyword cannot be used with constants.
+
+## Associated Constants
+
+Associated constants are constants associated with a type.
+
+```sway
+const ID: u32 = 0;
+```
+
```
Comment (otrho) [inline docs/book/src/basics/consts.md:16]: This is still a little be general, isn't it?  Associated consts are only declared and/or implemented in (i.e., associated with) traits and impls, yeah?  The above is just a regular non-associated const.  I think that code snippet should go in the previous `#Constants` section.

**Review by otrho — COMMENTED** [review]
Apart from extra doc nit, LGTM.

File: docs/reference/src/documentation/language/built-ins/constants.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# Constants
+
+A constant in Sway is an `associated item` which is an item associated to a particular struct type.
```
Comment (mohammadfawaz) [inline docs/reference/src/documentation/language/built-ins/constants.md:3]: This sentence seems misleading right? Constants are more than just "associated constants for struct types"

File: sway-core/src/control_flow_analysis/flow_graph/namespace.rs [hunk 5]
```diff
@@ -81,10 +81,26 @@ impl ControlFlowNamespace {
         self.function_namespace
             .insert((ident, TyFunctionSig::from_fn_decl(fn_decl)), entry);
     }
-    pub(crate) fn get_constant(&self, ident: &Ident) -> Option<&NodeIndex> {
+    pub(crate) fn get_constant(&self, const_decl: &TyConstantDecl) -> Option<&NodeIndex> {
+        self.const_namespace.get(&const_decl.name().clone())
+    }
+    #[allow(dead_code)]
```
Comment (mohammadfawaz) [inline sway-core/src/control_flow_analysis/flow_graph/namespace.rs:87]: How come this is dead?

File: docs/book/src/basics/consts.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+Alternatively, you can also _declare_ it in the trait, and implement it in the interface of the
+types implementing the trait.
+
+```sway
+script;
+
+trait ConstantId {
+    const ID: u32;
```
[the remark this answers is not in the document]
Comment (mohammadfawaz) [reply docs/book/src/basics/consts.md:43]: I guess what I'm saying is that I'm in favor of the type ascription being required in this situation.

**Review by mohammadfawaz — COMMENTED** [review]
Great! Looks good.

**Review by esdrubal — APPROVED** [review]

**Pull request merged**